src/kafka-util/src/bin/kgen.rs

```diff
@@ -678,48 +678,74 @@ async fn main() -> anyhow::Result<()> {
-            key_buf.clear();
-            key_buf.extend(key_dist.sample(&mut rng).to_be_bytes().iter())
-        } else {
-            key_buf.clear();
-            key_buf.extend(u64::cast_from(i).to_be_bytes().iter())
-        };
+    let counter = AtomicUsize::new(0);
+    thread::scope(|scope| {
+        for thread in 0..threads {
+            let counter = &counter;
+            let topic = &args.topic;
+            let mut key_gen = key_gen.clone();
+            let mut value_gen = value_gen.clone();
+            let producer: ThreadedProducer<DefaultProducerContext> = ClientConfig::new()
+                .set("bootstrap.servers", args.bootstrap_server.to_string())
+                .create()
+                .unwrap();
+            let mut key_buf = vec![];
+            let mut value_buf = vec![];
+            let mut n = args.num_records / threads;
+            if thread < args.num_records % threads {
+                n += 1;
+            }
+            scope.spawn(move |_| {
+                let mut rng = thread_rng();
+                for _ in 0..n {
+                    let i = counter.fetch_add(1, Ordering::Relaxed);
+                    if !args.quiet && i % 100_000 == 0 {
+                        eprintln!("Generating message {}", i);
+                    }
+                    value_gen.next_value(&mut value_buf, &mut rng);
+                    if let Some(key_gen) = key_gen.as_mut() {
+                        key_gen.next_value(&mut key_buf, &mut rng);
+                    } else if let Some(key_dist) = key_dist.as_ref() {
+                        key_buf.clear();
+                        key_buf.extend(key_dist.sample(&mut rng).to_be_bytes().iter())
+                    } else {
+                        key_buf.clear();
+                        key_buf.extend(u64::cast_from(i).to_be_bytes().iter())
+                    };
 
-        let mut rec = BaseRecord::to(&args.topic)
-            .key(&key_buf)
-            .payload(&value_buf);
-        if args.partitions_round_robin != 0 {
-            rec = rec.partition((i % args.partitions_round_robin) as i32);
-        }
+                    let mut rec = BaseRecord::to(&topic).key(&key_buf).payload(&value_buf);
+                    if args.partitions_round_robin != 0 {
+                        rec = rec.partition((i % args.partitions_round_robin) as i32);
+                    }
+                    let mut rec = Some(rec);
 
-        loop {
-            match producer.send(rec) {
-                Ok(()) => break,
-                Err((KafkaError::MessageProduction(RDKafkaErrorCode::QueueFull), rec2)) => {
-                    rec = rec2;
-                    thread::sleep(Duration::from_secs(1));
-                }
-                Err((e, _)) => {
-                    return Err(e.into());
+                    Retry::default()
+                        .clamp_backoff(Duration::from_secs(1))
+                        .retry(|_| match producer.send(rec.take().unwrap()) {
```

Comment:
This `unwrap` will cause a panic when we get a non-"QueueFull" error, because we will retry without resetting `rec` (in the previous code, we would return the error in that case).

That's probably fine because the behavior in either case is the program failing, but if we do choose to have that behavior, we should make it more explicit (e.g., panic on receiving the error, rather than here).

Reply:
Oh good point! I'll address.